io-engine/src/replica_backend.rs

```diff
@@ -18,39 +17,121 @@ pub trait ReplicaOps: Share + LogicalVolume {
+        Ok(share)
     }
 
+    fn shared(&self) -> Option<Protocol>;
+    fn create_ptpl(
+        &self,
+    ) -> Result<Option<PtplProps>, crate::pool_backend::Error>;
+
     /// Shares the replica via nvmf.
     async fn share_nvmf(
         &mut self,
         props: crate::core::NvmfShareProps,
-    ) -> Result<String, Self::ReplError>;
+    ) -> Result<String, crate::pool_backend::Error>;
     /// Unshare the replica.
-    async fn unshare(&mut self) -> Result<(), Self::ReplError>;
+    async fn unshare(&mut self) -> Result<(), crate::pool_backend::Error>;
     /// Update share properties of a currently shared replica.
-    async fn update_properties<P: Into<UpdateProps>>(
+    async fn update_properties(
         &mut self,
-        props: P,
-    ) -> Result<(), Self::ReplError>;
+        props: UpdateProps,
+    ) -> Result<(), crate::pool_backend::Error>;
 
     /// Resize the replica to the given new size.
-    async fn resize(&mut self, size: u64) -> Result<(), Self::ReplError>;
+    async fn resize(
+        &mut self,
+        size: u64,
+    ) -> Result<(), crate::pool_backend::Error>;
     /// Set the replica's entity id.
     async fn set_entity_id(
         &mut self,
         id: String,
-    ) -> Result<(), Self::ReplError>;
+    ) -> Result<(), crate::pool_backend::Error>;
     /// Destroy the replica from its parent pool.
     /// # Warning
     /// Destroying implies unsharing, which might fail for some reason, example
     /// if the target is in a bad state, or if IOs are stuck.
     /// todo: return back `Self` in case of an error.
-    async fn destroy(self) -> Result<(), Self::ReplError>;
+    async fn destroy(self: Box<Self>)
+        -> Result<(), crate::pool_backend::Error>;
+}
+
+/// Find replica with filters.
+#[derive(Debug, Default)]
+pub struct ListReplicaArgs {
+    /// Match the given replica name.
+    pub name: Option<String>,
+    /// Match the given replica uuid.
+    pub uuid: Option<String>,
+    /// Match the given pool name.
+    pub pool_name: Option<String>,
+    /// Match the given pool uuid.
+    pub pool_uuid: Option<String>,
+}
+
+/// Find replica with filters.
+#[derive(Debug, Clone)]
+pub struct FindReplicaArgs {
+    /// The replica uuid to find for.
+    pub uuid: String,
+}
+impl FindReplicaArgs {
+    /// Create `Self` with the replica uuid.
+    pub fn new(uuid: &str) -> Self {
+        Self {
+            uuid: uuid.to_string(),
+        }
+    }
+}
+
+/// Interface for a replica factory which can be used for various
+/// listing operations, for a specific backend type.
+#[async_trait::async_trait(?Send)]
+pub trait ReplicaFactory<Replica, Error> {
+    async fn find(
+        &self,
+        args: &FindReplicaArgs,
+    ) -> Result<Option<Replica>, Error>;
+    async fn list(&self, args: &ListReplicaArgs)
+        -> Result<Vec<Replica>, Error>;
+    fn backend(&self) -> PoolBackend;
+}
+
+#[macro_export]
+macro_rules! repl_factory {
```

Comment:
I am confused here! Why are we implementing ReplicaFactory via macro here, and we have implemented it already as:
> impl ReplicaFactory<LogicalVolume, Error> for ReplLvmFactory
and
> impl ReplicaFactory<Lvol, LvsError> for ReplLvsFactory

Reply:
This was to automatically generate a factory with the grpc types, example: `PoolFactory<Pool, tonic::Status>`
I have now made the trait more generic, and grpc conversion implemented without a trait so these macros will all be removed.